File: sd/slcanx-cpp/examples/01_simple_std.cpp
#include "slcanx.hpp"
#include <iostream>
#include <thread>
#include <chrono>

using namespace slcanx;

int main(int argc, char** argv) {
    std::string port = "COM3";
    if (argc > 1) port = argv[1];

    std::cout << "Opening " << port << " Channel 0 @ 500000bps" << std::endl;

    Slcanx slcan(port);
    
    slcan.set_rx_callback([](uint8_t ch, const CanFrame& frame) {
        if (ch == 0) {
            std::cout << "Rx: ID=" << std::hex << frame.id << " DLC=" << frame.data.size() << std::dec << std::endl;
        }
    });

    slcan.close_channel(0);
    slcan.set_bitrate(0, 500000);
    slcan.open_channel(0);

    std::vector<uint8_t> data = {0x11, 0x22, 0x33, 0x44};
    CanFrame frame = CanFrame::new_std(0x123, data);
    
    std::cout << "Sending frame..." << std::endl;
    slcan.send(0, frame);

    std::cout << "Listening... (Ctrl+C to exit)" << std::endl;
    while(true) {
        std::this_thread::sleep_for(std::chrono::seconds(1));
    }

    return 0;
}

File: sd/slcanx-rs/examples/01_simple_std.rs
use clap::Parser;
use slcanx::{Slcanx, CanFrame};
use std::time::Duration;
use anyhow::Result;

#[derive(Parser)]
struct Args {
    #[arg(short, long, default_value = "COM3")]
    port: String,
    #[arg(short, long, default_value_t = 500000)]
    bitrate: u32,
    #[arg(short, long, default_value_t = 0)]
    channel: u8,
}

fn main() -> Result<()> {
    env_logger::init();
    let args = Args::parse();

    println!("Opening {} Channel {} @ {}bps", args.port, args.channel, args.bitrate);

    // Grouping window 125us
    let slcan = Slcanx::new(&args.port, 115200, Duration::from_micros(125))?;
    let channel = slcan.get_channel(args.channel);

    channel.close()?;
    channel.set_bitrate(args.bitrate)?;
    channel.open()?;

    // Send a frame
    let frame = CanFrame::new_std(0x123, &[0x11, 0x22, 0x33, 0x44]);
    println!("Sending: {:?}", frame);
    channel.send(frame)?;

    println!("Listening... (Ctrl+C to exit)");
    loop {
        if let Some((ch, msg)) = slcan.recv_timeout(Duration::from_secs(1))? {
            if ch == args.channel {
                println!("Received: {:?}", msg);
            }
        }
    }
}
